### PR TITLE
Lexer & Parser

### PR DESCRIPTION
**Pull request by Negrini085**
So far we have defined the scenes to be rendered directly in the source code or otherwise in .nim files. This working methodology is clearly not appropriate for code that is intended to be usable by users who are not experts in the Nim programming language. A better (and also more convenient) way of defining the geometric shapes constituting the scenes is to implement a compiler that can interpret text files in which, by means of a specific syntax, there is all the information needed to perform the task previously introduced.

We need to implement:

- [x] a Lexer
- [x] a Parser

The lexer breaks down the source code into simple elements, called tokens, and reports lexical errors, while the parser analyzes the sequence of tokens to bind them together and understand their syntax and semantics.

To make the lexer fit the task we want to assign to it, we need to add the following functionality to the code:

- [x] Source location (in order to store where the particular token is read)
- [x] Grammar error (in order to raise error message whenever is needed)
- [x] Input stream (in order to be able to read input file)
- [x] Token (we need to define all the needed token in order to have a functional lexer)
- [x] Read token (a procedure that actually reads tokens)

We now have to add a parser to be able to actually create a scene from a text file. In order to do so, we need to add:

- [x] Expect procedures (sometimes we just need a specific token at a specific location)
- [x] Parse procedures (here we actually have to be able to create variables from text)
- [x] Change main in order to be able to actually read files

### CODE REVIEW

**Comment by Negrini085** [comment]
I feel like something is wrong with tabs

**Comment by Negrini085** [comment]
Be careful about:
```nim
var dSc
```
as ```expectNumber``` function argument

**Comment by Negrini085** [comment]
I feel like we should discuss about lexer and parser implementation and what we want to do. Right now i am not letting the user to have the possibility to chose treeKind and such stuff, because i would like to make this code usable by anyone (so they don't necessary have to know BVH and this kind of stuff), but i feel like we could devise a better strategy. Right now i would like to focus your attention on:

-  how to chose tkBinary, ...
-  how to chose maxShapesPerLeaf
-  how to give input parameters for pcg random generator

**Comment by Negrini085** [comment]
Also here I'm not really sure what to do. It could be useful to define a parseRenderer proc to define the renderer in the scene file, but i don't like it because in order to have a different renderer you would need to modify the scene defining file. Right now, I'm giving a default OnOff rederer as camera.renderer

**Comment by Negrini085** [comment]
Would it be better to add reflectance and treshold angle evaluation directly from parsing?

**Comment by Negrini085** [comment]
I would delete demo example, it should be useless

**Comment by Negrini085** [comment]
I'm going to delete demo task and demo.nim file in examples, it's just useless considering that we now have a demo.txt file

**Comment by Negrini085** [comment]
Be careful:

If you use:
```nim
let comp2 = newComposition(newRotX(90), newScaling(3))
```

Instead, if you type:
```nim
let comp2 = newComposition(newRotX(90), newScaling(3.0))
```

it's all good. I found it when i tried to modify a mesh example.

File: PhotoNim.nim [hunk 1]
```diff
@@ -1,10 +1,12 @@
 let PhotoNimVersion* = "PhotoNim 0.2"
 
-import src/[geometry, pcg, hdrimage, scene, material, hitrecord, camera]
-export geometry, pcg, hdrimage, scene, material, hitrecord, camera
+import src/[geometry, pcg, hdrimage, scene, material, hitrecord, camera, sceneFiles]
```
Comment (lorenzoliuzzo) [inline PhotoNim.nim:3]: I would better arrange code in the src/ directory in v0.3.1

Comment (Negrini085) [reply PhotoNim.nim:3]: Usted ya sabe

File: PhotoNim.nimble [hunk 2]
```diff
@@ -18,31 +18,42 @@ task build, "Build the `PhotoNim` executable\n":
...
 
-    for i in 0..paramCount(): commands[i] = paramStr(i)
-    for i in commands.find("demo")..paramCount(): demoCommand.add(" " & paramStr(i))
+    for i in 0..paramCount(): commands.add paramStr(i)
 
-    exec demoCommand
-    
+    for i in commands.find("demo")..<commands.len: 
+
+      if i == commands.find("demo"):
```
[the remark this answers is not in the document]
Comment (Negrini085) [reply PhotoNim.nimble:37]: I know, but i would like to let the user to have only the possibility to chose the renderer kind and to still call it:
```
nimble demo rendererKind
```

**Pull request merged**